streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -781,6 +781,16 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
   )(implicit trace: ZTraceElement): ZChannel[Env1, InErr, InElem, InDone, OutErr2, OutElem2, OutDone2] =
     ZChannel.PipeTo(() => self, () => that)
 
+  def pipeToOrFail[Env1 <: Env, OutErr1 >: OutErr, OutElem2, OutDone2](
+    that: => ZChannel[Env1, Nothing, OutElem, OutDone, OutErr1, OutElem2, OutDone2]
+  )(implicit trace: ZTraceElement): ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone2] = {
+    final case class ChannelFailure(err: OutErr) extends Throwable
```

Comment:
There is a missing primitive here that combines two channels to produce a new channel that can accept either of their inputs and produce either of their outputs.